File: Editorial/1062_Longest_Repeating_Substring/Python3/2_Brute_Force_with_Incremental_Search.py
class Solution:
    def longestRepeatingSubstring(self, s: str) -> int:
        length = len(s)
        max_length = 0
        seen_substrings = set()

        start = 0
        while start < length:
            end = start
            # Stop if it's not possible to find a
            # longer repeating substring
            if end + max_length >= length:
                return max_length
            # Generate substrings of length max_length + 1
            current_substring = s[end : end + max_length + 1]
            # If a repeating substring is found,
            #  increase max_length and restart
            if current_substring in seen_substrings:
                start = -1  # Restart search for new length
                seen_substrings.clear()
                max_length += 1
            else:
                seen_substrings.add(current_substring)
            start += 1

        return max_length

File: Editorial/1062_Longest_Repeating_Substring/CPP/2_Brute_Force_with_Incremental_Search.cpp
class Solution {
public:
    int longestRepeatingSubstring(string s) {
        int length = s.length(), maxLength = 0;
        unordered_set<string> seenSubstrings;

        for (int start = 0; start < length; start++) {
            int end = start;
            // Stop if it's not possible to find a longer repeating substring
            if (end + maxLength >= length) {
                return maxLength;
            }
            // Generate substrings of length maxLength + 1
            string currentSubstring = s.substr(end, maxLength + 1);
            // If a repeating substring is found, increase maxLength and
            // restart
            if (!seenSubstrings.insert(currentSubstring).second) {
                start = -1;  // Restart search for new length
                seenSubstrings.clear();
                maxLength++;
            }
        }
        return maxLength;
    }
};
